Parallelise the testsuite in dev (as possible) and in CI

nicoonoclaste opened the pull request:
### Changes

- [x] `test.sh` checks whether `pytest-xdist` is available, and if so runs the test suite in parallel
- [x] `pytest-xdist` is installed in CI's test environments
- [x] Introduce a new `requirements-ci.txt` for those environments, which cascades to `requirements-tests.txt`

### Motivation

Our test suite is getting slower, as it grows deeper and our API grows wider.  While unsurprising, this is an issue in 2 ways:

1. During (local) development, `test.sh` taking 8s is enough of a disruption that developers may subconsciously skip it, or specifically run a subset of the test suite, potentially missing issues which are only caught in CI once pushed, and getting much slower feedback.
2. In CI, `test.sh` takes about 1'30" (on Linux) which is long-enough one may skip to a different task etc., introducing further delays in the feedback loop, and more cognitive overhead from repeated context switching.  Ideally, I'd like our CI to complete within 10s of seconds (at least on PRs themselves, Bors may take longer as it's async)

Parallelising the test suite addresses both usecases:
1. `test.sh` got 4× faster in local development — 8s to 2s — fast-enough I could consider running it in a commit hook or such.
2. In the CI profile, the speedup was close to 8× — from 1'23" to 11" — which achieves my goal of getting feedback in tens of seconds.

nicoonoclaste committed:
test.sh: Autoparallelise the testsuite as possible

On my workstation, this yields a 4× reduction in test time in the dev profile,
and almost 8× under the CI test profile.  The difference is that Hypothesis run
each test 10 times (w/ different random inputs) in dev and 1k time in CI.

nicoonoclaste committed:
CI/Tests: Ensure pytest-xdist is available

nicoonoclaste commented:
Apparently, `pytest-xdist` is chocking on `@pytest.mark.parameterize`d tests which seem to be instanciated in non-deterministic order; I suspect the source of non-determinism might be the traversal order of `frozenset`s defined in `utils.py`.

nicoonoclaste committed:
tests: Determinize the traversal order of `*_OPS`

nicoonoclaste commented:
CI failures seems to be down to Cirrus' VMs being unusually slow, maybe due to noisy neighbours or such.

nicoonoclaste commented:
bors try

AstraLuma reviewed (approved):
That's some high standards. While waiting for mac is the much bigger problem, this is an improvement with minimal maintenance burden.

bors r+

nicoonoclaste commented:
bors r=astraluma

nicoonoclaste committed:
test.sh: Limit the degree of parallelism in Linux CI

Apparently, pytext-xdist wrongly detects there being 30 cores available (what
the host's CPU supports) rather than the default 2 available to the container
by default.

nicoonoclaste commented:
*grumbles*  Of course, requesting 8 CPUs causes the task to take forever before being scheduled  >_>'

On the plus side, I found the root cause of the flakyness issue: `pytest-xdist` was grabbing 30 threads, one for each logical core of the host, even though it was running in a container limited to 2 cores  ^^'

nicoonoclaste committed:
tests: Don't expect consistent test timings in CI

nicoonoclaste committed:
CI: Use “greedy containers” for Linux tests, disable CPU limit

This should let us get fast task scheduling while still taking advantage of
Cirrus' idle CPUs on the “community cluster”.

nicoonoclaste commented:
CPU usage literally off the charts ^^'

![2021-11-02_off_the_charts](https://user-images.githubusercontent.com/1155801/139684703-af22c2f0-7c7f-4467-8470-83e2369edb76.png)

nicoonoclaste commented:
Apparently, spinning up 30 PyPy processes does cause jobs to OoM, though  >_>'

nicoonoclaste committed:
CI: Limit PyPy tests to 15 processes on Linux

Otherwise the test container simply runs out of memory.

nicoonoclaste commented:
@AstraLuma Care to re-review?  I feel like there was enough ~~flailing around~~ incremental fixing to warrant that  ^^

AstraLuma reviewed (approved):
lgtm

bors r+
